packages/@aws-cdk/aws-applicationsignals-alpha/lib/slo/slo-types.ts

```diff
@@ -0,0 +1,138 @@
+import { ISlo, PeriodBasedSloProps, RequestBasedSloProps } from './slo';
+import { ComparisonOperator, MetricType } from './constants';
+import { Goal } from './interval';
+import { PeriodBasedAppSignalsMetricProps, PeriodBasedCloudWatchMetricProps, RequestBasedAppSignalsMetricProps, RequestBasedCloudWatchMetricProps } from './metric';
+import { Resource } from 'aws-cdk-lib';
+import { Construct } from 'constructs';
+import * as applicationsignals from 'aws-cdk-lib/aws-applicationsginals';
+
+/**
+ * Base class for all SLO implementations
+ */
+export abstract class ServiceLevelObjective extends Resource implements ISlo {
+    public readonly arn: string;
+    public readonly name: string;
+    public readonly goal: Goal;
+
+    protected constructor(
+        scope: Construct,
+        id: string,
+        protected readonly props: PeriodBasedSloProps | RequestBasedSloProps
```

Comment:
We do not recommend using union types. The Props should be of the base type of props, and in general the children classes of this abstract class will forward the concrete props classes. Also add docs.